src/crimson/os/seastore/extentmap_manager/btree/extentmap_btree_node_impl.cc

```diff
@@ -0,0 +1,604 @@
+            return merge_entry_ertr::make_ready_future<ExtMapNodeRef>(
+               lo >= pivot ? replacement_r : replacement_l);
+          });
+        });
+      });
+    }
+  });
+}
+
+
+ExtMapInnerNode::internal_iterator_t
+ExtMapInnerNode::get_containing_child(objaddr_t lo)
+{
+  // TODO: binary search
+  for (auto i = begin(); i != end(); ++i) {
+    if (i.contains(lo))
+      return i;
+  }
+  ceph_assert(0 == "invalid");
+  return end();
+}
+
+std::ostream &ExtMapLeafNode::print_detail(std::ostream &out) const
+{
+  return out << ", size=" << get_size()
+	     << ", depth=" << depth;
+}
+
+ExtMapLeafNode::seek_lextent_ret
+ExtMapLeafNode::seek_lextent(Transaction &t, objaddr_t lo, extent_len_t len)
+{
+  logger().debug(
+    "ExtMapLeafNode::seek_lextent {}~{}", lo, len);
+  auto ret = extent_map_list_t();
+  auto [from, to] = get_leaf_entries(lo, len);
+  if (from == to && to != end())
+    ++to;
+  for (; from != to; ++from) {
+    auto val = (*from).get_val();
+    ret.emplace_back(
+      std::make_unique<Extent>(
+	(*from).get_key(),
+	val.laddr,
+	val.lextent_offset,
+	val.length));
+    logger().debug("ExtMapLeafNode::seek_lextent find {}~{}", lo, val.laddr);
+  }
+  return seek_lextent_ertr::make_ready_future<extent_map_list_t>(
+    std::move(ret));
+}
+
+ExtMapLeafNode::insert_ret
+ExtMapLeafNode::insert(Transaction &t, objaddr_t lo, lext_map_val_t val)
+{
+  ceph_assert(!at_max_capacity());
+  if (!is_pending()) {
+    auto mut = make_duplicate(t, this)->cast<ExtMapLeafNode>();
+    return mut->insert(t, lo, val);
+  }
+  auto insert_pt = lower_bound(lo);
+  journal_insert(insert_pt, lo, val, maybe_get_delta_buffer());
+
+  logger().debug(
+    "ExtMapLeafNode::insert: inserted {}->{} {} {}",
+    insert_pt.get_key(),
+    insert_pt.get_val().laddr,
+    insert_pt.get_val().lextent_offset,
+    insert_pt.get_val().length);
+  return insert_ertr::make_ready_future<ExtentRef>(
+    std::make_unique<Extent>(lo, val.laddr, val.lextent_offset, val.length));
+}
+
+ExtMapLeafNode::punch_lextent_ret
+ExtMapLeafNode::punch_lextent(Transaction &t, objaddr_t lo, extent_len_t len)
+{
+  if (!is_pending()) {
+    auto mut = make_duplicate(t, this)->cast<ExtMapLeafNode>();
+    return mut->punch_lextent(t, lo, len);
+  }
+  auto result_up = std::make_unique<extent_map_list_t>();
+  auto &ret = *result_up;
+  auto deflist = std::list<laddr_t>();
+  auto [punch_pt, punch_end] = get_leaf_entries(lo, len);
+  objaddr_t loend = lo + len;
+  for (; punch_pt != punch_end; ++punch_pt) {
+    if (punch_pt->get_key() >= loend)
+      break;
+    logger().debug("ExtMapLeafNode::punch_lextent {} {}", lo, len);
+    auto val = (*punch_pt).get_val();
+    if ( punch_pt->get_key() < lo) {
+      if (punch_pt->get_key() + val.length > loend) {
+        //split and deref middle
+        logger().debug("ExtMapLeafNode::punch_middle {} {}", lo, len);
+	uint32_t front = lo - punch_pt->get_key();
```

Comment:
I don't think I understand this case.  If we have an extent at 12k~4k and we overwrite at 13k~3k, presumably, we need to drop the original 4k and create a new one.  We don't want to leave an entry at 12k~1k pointing at the old 4k extent.  I think perhaps punch_lextent should be required to start and end on existing mapping boundaries, or you should remove and include *all* extents overlapping lo~len.

Reply:
assume there is a data extent whose size is 3 pages, if modify the middle page, we can allocate from lba one new page for it. and punch the existing one 3 pages mapping to make  3  mappings. the middle page map to the new allocate,  the first and third page keep the origin mapping to the old laddr.   or you think in this case we should allocate 3 new pages from lba ?

Reply:
(Sorry, messed up github formatting in my example, fixed)

I guess I'm just not clear on how this extent is meant to be used.  It definitely doesn't make sense to shrink a mapping that already exists as your code appears to do here since the extent underlying the mapping didn't change size.  I think the choice as to how to split up the mapping should probably live in the consumer of this interface, so this interface should probably just return and remove all extents that overlap the specified region.  The caller can then re-add whatever it wants the result to be.

Reply:
return overlap the specified region is understand,  just want to make sure, "remove" only for overlap the whole mapping region case, right?  if overlap only a part of the mapping, we shouldn't remove the overlap, right?

Reply:
@liu-chunmei It might clarify matters if you wrote down an example of how the layer above this implementation might actually handle a write which overlaps only partially an existing allocation.

Reply:
for example, if user update some bytes in an existing mapping data extent. if  we duplicate the data extent no matter how big the extent and how small the updated part, modify  the extent map map to the new laddr, then we needn't punch hole.  I am wondering if this is your thinking? At the beginning, the punch hole is refer bluestore punch hole, blue store is a little complex logic to use blob.

Reply:
I think that's probably an ok strategy for now.  Either way, however, I really don't think that logic belongs in the extent map.  Try to come up with a design that leaves that choice to the next layer up.

Reply:
then I should remove all punch hole code from extentmap tree, right?

Reply:
Well, you need some way of communicating to the interface user which extents exist, and a way of replacing one set of extents with another.  Perhaps one call which returns the set of extents overlapping a particular offset range and a second which specifies the new set of extents implicitly replacing any that overlap them?

```
future<set::list<extent>> get_extents(..., off_t offset, extent_len_t len);
future<> set_extents(..., std::list<extent> extents,  off_t offset, extent_len_t len);
```

What you choose to do here depends on how you intend for the interface to be used.  My point is basically that the choice of exactly when to
1) mutate existing extent
2) split extent
will depend on a lot of details and is likely to get more complicated in the future, so we don't want the logic for making that choice buried down in the logic for manipulating the extent map btree.

Reply:
"one call which returns the set of extents overlapping a particular offset range"   ------- I think find_lextent will do this work. 
"second which specifies the new set of extents implicitly replacing any that overlap them" ---- I will add mutate_lextent to replace the whole extent mapping. 
"split extent" ---what do you mean split here?  the user can remove existing mapping and add two news, right?  or you want do different thing?

Reply:
Yes, that is what I mean by split extent.

Reply:
just make sure, mutate_lextent only replace mapped Value (laddr), object offset and length will keep the same.  That means, when update some part in existing data extent, we duplicate the whole extent, and remapping to the duplicated extent, right? 
for split extent, there is already rm_lextent and add_lextent interface.  do you mean we add a new interface to do rm and add in one call?  in what cases we split extent?

Reply:
mutate_lextent seems too restrictive, the caller might want to replace a single extent with several extents.  rm_lextent and add_lextent would be sufficient, but I anticipate that you'd end up traversing the tree multiple times for the same IO in many cases.  I think the user's logic probably looks something like this:

1) interpret IO
2) get the set of extents relevant to that IO (should just be all current extents within a range, something like extent_set find_lextents(..., offset, len))
3) submit the new extents for that range.  These extents would be a combination of existing extents where they don't need to change and new extents where they do.  They'd then submit that whole set at once to avoid having to traverse the tree once for each extent added or removed.

This structure enables the user to have complete control over policy for when to mutate vs replace an existing extent while ensuring that we only need to access the tree twice.

Reply:
As you modify your BtreeExtentMapManager, make sure to add comments specifying exactly what the methods do.

Reply:
> As you modify your BtreeExtentMapManager, make sure to add comments specifying exactly what the methods do.

will update it

Reply:
> mutate_lextent seems too restrictive, the caller might want to replace a single extent with several extents. rm_lextent and add_lextent would be sufficient, but I anticipate that you'd end up traversing the tree multiple times for the same IO in many cases. I think the user's logic probably looks something like this:
> 
> 1. interpret IO
> 2. get the set of extents relevant to that IO (should just be all current extents within a range, something like extent_set find_lextents(..., offset, len))
----- I think current find_lextent will return a extent list for a certain (offset len) pair,  do we need add an interface for a list  of (offset len) pair?   for each io, only one offset and len, right? 

> 3. submit the new extents for that range.  These extents would be a combination of existing extents where they don't need to change and new extents where they do.  They'd then submit that whole set at once to avoid having to traverse the tree once for each extent added or removed.
----for the combined extents, they may distribute in different leaves, so for each extent, we will go through  the tree from root node , right? so the combined extents maybe keep it as before, or  just update the mapped val part, or remove old one and add several news. then the logical in extenmap will be complex, how extentmap know which part need update, which one need remove? 
> 
> This structure enables the user to have complete control over policy for when to mutate vs replace an existing extent while ensuring that we only need to access the tree twice.

Reply:
I don't see a find_lextent method, debug methods in find_hole mention it though?

If you think add/remove would be simpler, that's fine too.

Reply:
To bring this back to my initial comment, I'm still just unclear about what find_hole is meant to do.  It may be that adding a more detailed explanation could solve it, but there are some things aren't obvious to me:
- I don't see a case where the interface user would need to find an unmapped extent (hole)
- I don't understand ExtMapLeafNode::find_hole:
  - Why is it valid to read the extent at at start->get_val() as a TestBlock (that extent generally won't be a TestBlock)?
  - Why would this component need to read the mapped block?
  - I don't really understand the L_ADDR_NULL returns early in the method

Reply:
Actually, is ExtMapLeafNode::find_hole just reading those extents in order to get the length of the block?

Reply:
I think punch_lextent is a problem because reducing the size of an extent should probably be done in the calling logic -- this component should simply manage the offset mapping.

Reply:
> I don't see a find_lextent method, debug methods in find_hole mention it though?
> 
> If you think add/remove would be simpler, that's fine too -- as I said, that should be sufficient.

the old name is seek_lextent, you suggested to change it to find_lextent, I will update code later.

Reply:
Oh, by find_lextents you meant seek_lextents from my above comment, I think.  Yeah, that seems appropriate.

Reply:
find hole:  I consider this case, for example, there is a map obj offset -> laddr ( 0-> 0x 1000) length 0x200,  and lab allocate at least one page for this data block, then next write to obj offset 0x200 length 0x300, find_lextent can't return the 0->0x1000 map, we have two choice, ask lba allocate a new data page for obj offset 0x200 , another choice to get the first one and duplicate and  they share the same page,  find hole try to do the second choice. 
the length in extent map leaf node is obj offset and length, not lba allocate LogicalExtenCache length, I should know how big the LogicalExtentCache to figure out if there is space left.  I can't get LogicalExtentCache  length.

Reply:
Oh, you are trying to find *unused* space in an existing mapping.  I don't think that kind of unmapped space should be possible in the first place -- if part of an extent becomes replaced, the caller should have dropped the extent completely and replaced it.

Reply:
That is, the logical extent length should always match the underlying lba extent length.

Reply:
For small objects/overwrites, you should probably assume that offset and length will be block aligned -- the caller can handle overlaying small objects/writes into 4k extents.

Reply:
you mean the key is block aligned in extent map tree?

Reply:
the length shouldn't be block aligned, we should know exactly how long data stored.

Reply:
No, the caller can simply map a 4k extent at 0 and the onode can track the actual size.

Reply:
I'm open to a more complicated scheme later on, but this is simple enough for now.

Reply:
> No, the caller can simply map a 4k extent at 0 and the onode can track the actual size.

I don't understand it. currently extent map  map obj offset -> laddr, lextent_offset( data start offset in this laddr extent), length ( real data length for this map), offset can be block aligned, since we can use lextent_offset to calculate it.

Reply:
> I'm open to a more complicated scheme later on, but this is simple enough for now.

I will remove find_hole and punch_hole, and add mutate_lextent to update the mapped value otherwise need remove first and add  two times go through the tree.

Reply:
I don't think you need the lextent_offset because we can choose for it to be always 0.  Suppose the user writes 100~200 to an empty object.  Seastore (above this interface) would create a new onode, allocate an extent of size 4k, add that 4k extent at offset 0 to the extent map, and write the 100 bytes at offset 100 in the extent.  The onode itself could then remember that the object has 4k bytes allocated, but has size 200.

There are reasons to use a subset of an extent, but a lot of other design would have to happen first.

Reply:
I think for efficiency, we'll eventually need something more powerful than mutate_lextent, perhaps add and remove would suffice for now.